Add explicit collectives grouping pass.

chaserileyroberts opened the pull request:
This PR adds an explicit frontend attribute to jitted JAX methods that will force the computation to be run in a single NCCL Group. 

The intention of this change is to enable "multi-directional communications" that better saturate NVLink systems. We reuse the existing NCCLGroupThunk logic during the IR emitter stage, so we only need to introduce the async wrapper and call inliner.

Here is an example of using this feature from JAX.

```python
import jax
from jax._src.xla_metadata import set_xla_metadata
from jax.experimental.shard_map import shard_map
from jax.sharding import Mesh, PartitionSpec as P
from functools import partial

num_devices = 4
mesh = Mesh(np.array(jax.devices()), ('i',))

# Unique source target pairs for our two ppermutes later.
perm_up = [(i, (i+1) % num_devices) for i in range(num_devices)]
perm_down = [(i, (i-1) % num_devices) for i in range(num_devices)]

# NCCL Group computations _must_ be jitted.
@jax.jit
def bidir_comms(a):
    b = jax.lax.ppermute(a, "i", perm_up)
    c = jax.lax.ppermute(a, "i", perm_down)
    return b, c

@jax.jit
@partial(shard_map, mesh=mesh, in_specs=P(None, 'i'), out_specs=P(None, 'i'))
def groups(a):
   # Running our jitted function in this context will force the use of a NCCL Group.
    with set_xla_metadata(_collectives_group="", inlineable="false"):
        b, c = bidir_comms(a)
    return b + c
```

This is nsys trace with the annotation.
![Screenshot 2025-03-16 at 8 32 44 PM](https://github.com/user-attachments/assets/4f9ea41b-5b78-40ab-a1f1-c58815c963de)


Vs without the additional annotation
![Screenshot 2025-03-16 at 8 31 56 PM](https://github.com/user-attachments/assets/4937107e-f457-424e-8e36-73512b13dc5c)

As you can see, there is only a single NCCL kernel in the annotated example.

rosiezou commented on xla/service/hlo_verifier.cc:
```diff
@@ -3104,13 +3092,6 @@ absl::StatusOr<bool> HloVerifier::Run(
-      // collection of send/recv instructions. This is needed to represent NCCL
-      // groups on GPU.
-      if (computation->IsAsyncComputation() &&
-          !computation->OnlyContainsSendRecv()) {
```
This part of HLO verifier is not GPU-specific. Instead of deleting the check, I suggest the following:
1. Keep `VerifyAsyncComputation`
2. Change the condition from `computation->IsAsyncComputation() && !computation->OnlyContainsSendRecv()` to the following
```
// in hlo_computation.h and hlo_computation.cc, add a check for caller type, something that looks like this
bool IncludedInAsyncNcclGroup(){
    std::unique_ptr<CallGraph> call_graph = CallGraph::Build(this->parent_);
    std::vector<HloInstruction*> callers = call_graph->GetComputationCallers(computation);
    for (HloInstruction* caller : callers) {
        if (caller->opcode() == HloOpcode::kCall && caller->get_frontend_attribute(kNcclGroupAttr) != nullptr){
            return true;
        }
    return false;
}

// In hlo_verifier.cc, line 3110
if (computation->IsAsyncComputation &&
    !computation->OnlyContainsSendRecv() &&
    !IncludedInAsyncNcclGroup()) {
    TF_RETURN_IF_ERROR(VerifyAsyncComputation(computation));
}
```
This way we don't break the parts of XLA that requires async-wrapped computations to only contain a single operation.

jreiffers replied:
Creating a call graph should no longer be necessary here, computations now know their callers: https://github.com/openxla/xla/blob/7639178ebd11a1248a2fb3ac9138c1a14f0353aa/xla/hlo/ir/hlo_computation.h#L960

chaserileyroberts replied:
I think just adding a check if the async op has the `_nccl_group` frontend attribute would be sufficient.

rosiezou replied:
Yea that works too. I'm cool with whatever implementation as long as you're not skipping the check completely.

chaserileyroberts replied:
I modified the check and put back all of the tests I removed originally.

rosiezou commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper.cc:
```diff
@@ -0,0 +1,77 @@
+  bool changed = false;
+  for (const HloComputation* comp : module->computations()) {
+    for (HloInstruction* instr : comp->instructions()) {
+      TF_ASSIGN_OR_RETURN(bool result, CreateNcclGroupAsyncPair(instr));
```
Looks like your feature supports creating more than one nccl group per program. If that is the case, please add corresponding test cases that contain multiple nccl groups.

chaserileyroberts replied:
This was a good catch. Previously, if you used the same computation twice for two different async nccl groups, then this [check](https://github.com/openxla/xla/blob/a31cce0bcfa13aedf6ca2ba398eea8781c014e1d/xla/hlo/ir/hlo_instructions.cc#L405) fails during the construction of the second async start instruction.

I modified the code to instead create a clone of the computation and added tests for that situation.

rosiezou commented on xla/service/call_inliner.cc:
```diff
@@ -176,6 +176,10 @@ bool InlineComposites(
 }
 
+bool InlineNcclGroups(HloInstruction* instruction) {
+  return !instruction->frontend_attributes().map().contains(kNcclGroupAttr);
```
I believe this function & the corresponding test case in `call_inliner_test.cc` are unnecessary since the runtime behavior in XLA:GPU is the same for inlined vs. un-inlined functions. Both are essentially inlined for buffer assignment purpose and resulting `ThunkSequence` is the same.

chaserileyroberts replied:
We do end up needing this, but for a different reason than you state. 

The call operation is how we define the nccl grouping boundaries. If we don't include it in the inline exceptions, then the call will disappear very early in the compiler stack and we won't have a clear way to define our NCCL group during the async wrapper pass. Ultimately, the call operation does get deleted during the async wrapper pass, so it won't survive all the way to the thunk executions.

rosiezou replied:
Got it. Thanks for the clarification!

rosiezou reviewed (commented):
Left some comments. In summary:
1. Please clarify the behavior of the HLO pass and the expected dev ex for JAX users. In particular, please add examples for multiple NCCL groups if it is allowed, or clarify via inline comments and doc strings if you're limiting kCall nccl groups to a single group per program.
2. The way we test `ir_emitter_unnested.cc` is via execution tests such as `collective_ops_test.cc`. Please add similar execution tests for this feature. I have also left additional comments in the HLO pass regarding test coverage.

chaserileyroberts commented:
> Please clarify the behavior of the HLO pass and the expected dev ex for JAX users. In particular, please add examples for multiple NCCL groups if it is allowed, or clarify via inline comments and doc strings if you're limiting kCall nccl groups to a single group per program.

I added an example that uses multiple nccl groups from the same original computation and modified the program to make it work.

> The way we test ir_emitter_unnested.cc is via execution tests such as collective_ops_test.cc. Please add similar execution tests for this feature. I have also left additional comments in the HLO pass regarding test coverage.

There conveniently already exists `nccl_group_execution_test.cc` so I added a test there targeting the specific bidirectional communication pattern we want to eventually use.

chaserileyroberts commented:
@rosiezou This is ready for another pass whenever you have time.

rosiezou commented:
Please fix the CI failures caused by `call_inliner_test`. Other than that, LGTM.

chaserileyroberts commented:
> Please fix the CI failures caused by call_inliner_test. Other than that, LGTM.

I actually ended up deleting everything related to the `call_inliner`. Instead, the frontend attribute `inlineable` does what we need. I added that to the example code.

rosiezou commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper.h:
```diff
@@ -0,0 +1,48 @@
+==============================================================================*/
+
+#ifndef XLA_SERVICE_GPU_TRANSFORMS_NCCL_GROUP_ASYNC_WRAPPER_H_
+#define XLA_SERVICE_GPU_TRANSFORMS_NCCL_GROUP_ASYNC_WRAPPER_H_
```
@chaserileyroberts please use `TENSORFLOW_COMPILER_XLA_SERVICE_GPU_TRANSFORMS_EXPLICIT_NCCL_GROUP_ASYNC_WRAPPER_H_` in the header guard

chaserileyroberts replied:
Done

rosiezou commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper.h:
```diff
@@ -0,0 +1,48 @@
+
+}  // namespace xla::gpu
+
+#endif  // XLA_SERVICE_GPU_TRANSFORMS_NCCL_GROUP_ASYNC_WRAPPER_H_
```
same here please use `TENSORFLOW_COMPILER_XLA_SERVICE_GPU_TRANSFORMS_EXPLICIT_NCCL_GROUP_ASYNC_WRAPPER_H_` in the inline comment

chaserileyroberts replied:
Done

rosiezou commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper.h:
```diff
@@ -0,0 +1,48 @@
+      const absl::flat_hash_set<absl::string_view>& execution_threads) override;
+};
+
+}  // namespace xla::gpu
```
nit
```
}  // end namespace gpu
}  // end namespace xla
```

chaserileyroberts replied:
done

chaserileyroberts commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper.h:
```diff
@@ -0,0 +1,48 @@
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/hlo/pass/hlo_pass_interface.h"
+
+namespace xla::gpu {
```
Done, here and in the .cc file.

ezhulenev commented on xla/service/gpu/transforms/explicit_nccl_group_async_wrapper_test.cc:
```diff
@@ -0,0 +1,106 @@
+namespace xla::gpu {
+namespace {
+
+using ExplicitNcclGroupAsyncWrapperTest = HloTestBase;
```
We removed Nccl from all type names in XLA:GPU and we use generic "collectives" + generic collectives APIs everywhere. Can you rename it to `ExplicitCollectiveGroupAsyncWrapper`.

chaserileyroberts replied:
Sure, but this pass might be meaningless on non-Nvidia hardware. We also are still using `_nccl_group` as the frontend attribute, do you think that needs to be changed too?

ezhulenev replied:
Yes, eventually most mentions of nccl should go away (when they have a generic replacement).

chaserileyroberts replied:
I'll change it to `_collectives_group`
